pkg/providers/securitygroup/securitygroup.go

```diff
@@ -25,37 +25,47 @@ import (
 	"github.com/mitchellh/hashstructure/v2"
 	"github.com/patrickmn/go-cache"
 	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/types"
 	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"sigs.k8s.io/karpenter/pkg/utils/pretty"
 
 	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
 )
 
 type Provider interface {
-	List(context.Context, *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error)
+	List(context.Context, *v1beta1.EC2NodeClass) ([]v1beta1.SecurityGroup, error)
+	UpdateSecurityGroups(context.Context, *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error)
 }
 
 type DefaultProvider struct {
-	sync.Mutex
-	ec2api ec2iface.EC2API
-	cache  *cache.Cache
-	cm     *pretty.ChangeMonitor
+	sync.RWMutex
+	ec2api     ec2iface.EC2API
+	kubeClient client.Client
+	cache      *cache.Cache
+	cm         *pretty.ChangeMonitor
 }
 
-func NewDefaultProvider(ec2api ec2iface.EC2API, cache *cache.Cache) *DefaultProvider {
+func NewDefaultProvider(kubeClient client.Client, ec2api ec2iface.EC2API, cache *cache.Cache) *DefaultProvider {
 	return &DefaultProvider{
-		ec2api: ec2api,
-		cm:     pretty.NewChangeMonitor(),
+		ec2api:     ec2api,
+		kubeClient: kubeClient,
+		cm:         pretty.NewChangeMonitor(),
 		// TODO: Remove cache cache when we utilize the security groups from the EC2NodeClass.status
 		cache: cache,
 	}
 }
 
-func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]*ec2.SecurityGroup, error) {
-	p.Lock()
-	defer p.Unlock()
+func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) ([]v1beta1.SecurityGroup, error) {
```

Comment:
So, rather than relying on the provider for list operations, why doesn't our launch path just pull security groups from the status? Then you don't rely on the providers in the launch path, you only rely on the providers

Reply:
That also means that your List() call doesn't have to change. From a conceptual perspective: List() on the provider continues to mean that you are listing security groups on the cloudprovider and then the status just always holds the info for the launch path and for drift